addons/web/controllers/json.py

```diff
@@ -0,0 +1,254 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import ast
+import functools
+import logging
+from collections import defaultdict
+from datetime import date
+from dateutil.relativedelta import relativedelta
+from http import HTTPStatus
+from lxml import etree
+from urllib.parse import urlencode
+
+from werkzeug.exceptions import BadRequest
+
+from odoo import http
+from odoo.exceptions import AccessError
+from odoo.http import request
+from odoo.models import regex_object_name
+from odoo.osv import expression
+from odoo.tools import lazy
+from odoo.tools.safe_eval import safe_eval
+from .utils import get_action_triples
+
+
+_logger = logging.getLogger(__name__)
+
+
+class WebJsonController(http.Controller):
+
+    @http.route('/json/<path:subpath>', auth='public', type='http', readonly=True)
+    def web_json(self, subpath, **kwargs):
+        return self.web_json_18_0(subpath, **kwargs, redirect=True)
```

Comment:
Why don't you simply return `request.redirect(...)` like before?
It adds complexity to the main method.

Reply:
Because we want to add relevant parameters to the URL.

Reply:
I make it this way to have 1 redirect instead of 2. Do you want me to roll it back?

Reply:
Thanks @rco-odoo to spot this out. I missed it out of the MOVE+IMP commit. Indeed the redirection is part of the API, if the query-string is lost during the redirection we can force it stay the same using `request.redirect_query(..., request.httprequest.args)`.

Would it be possible to split the commit in two? On `[MOVE]` that is only the result of isolating the existing code inside a dedicated file. And a `[IMP]` that adds the changes?

Reply:
My original message maybe lacked clarity. **It must redirect, always.** `/json/<action>` has no representation, it is a trick so users can easily change `/odoo/<action>` to `/json/<action>` in the web client's URL and get the most recent json representation of the data. In script people must precise the version, otherwise one day their script will crash because we would had change the redirection to v2. It should be it easy for people to copy/paste the URL with the version, and hard to copy/paste the on without.